process/sync/metablock.go

```diff
@@ -292,40 +306,45 @@ func (boot *MetaBootstrap) areNotarizedShardHeadersFound(
 }
 
 func (boot *MetaBootstrap) applyNotarizedBlocks(
-	finalNotarized map[uint32]uint64,
-	lastNotarized map[uint32]uint64,
+	finalNotarized map[uint32]*HdrInfo,
+	lastNotarized map[uint32]*HdrInfo,
 ) error {
 	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
-		nonce := finalNotarized[i]
-		if nonce > 0 {
-			headerHandler, _, err := boot.getShardHeaderFromStorage(i, nonce)
-			if err != nil {
-				return err
-			}
+		if finalNotarized[i] == nil || lastNotarized[i] == nil {
+			return ErrNilNotarizedHeader
+		}
+		if finalNotarized[i].Hash == nil || lastNotarized[i].Hash == nil {
+			return ErrNilHash
+		}
 
-			boot.blkExecutor.AddLastNotarizedHdr(i, headerHandler)
+		headerHandler, err := process.GetShardHeaderFromStorage(finalNotarized[i].Hash, boot.marshalizer, boot.store)
+		if err != nil {
+			return err
 		}
 
-		nonce = lastNotarized[i]
-		if nonce > 0 {
-			headerHandler, _, err := boot.getShardHeaderFromStorage(i, nonce)
-			if err != nil {
-				return err
-			}
+		boot.blkExecutor.AddLastNotarizedHdr(i, headerHandler)
 
-			boot.blkExecutor.AddLastNotarizedHdr(i, headerHandler)
+		headerHandler, err = process.GetShardHeaderFromStorage(lastNotarized[i].Hash, boot.marshalizer, boot.store)
+		if err != nil {
+			return err
 		}
+
+		boot.blkExecutor.AddLastNotarizedHdr(i, headerHandler)
 	}
 
 	return nil
 }
 
-func (boot *MetaBootstrap) cleanupNotarizedStorage(lastNotarized map[uint32]uint64) {
+func (boot *MetaBootstrap) cleanupNotarizedStorage(lastNotarized map[uint32]*HdrInfo) {
 	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
 		hdrNonceHashDataUnit := dataRetriever.ShardHdrNonceHashDataUnit + dataRetriever.UnitType(i)
 		highestNonceInStorer := boot.computeHighestNonce(hdrNonceHashDataUnit)
 
-		for i := lastNotarized[i] + 1; i <= highestNonceInStorer; i++ {
+		if lastNotarized[i] == nil {
+			continue
+		}
+
+		for i := lastNotarized[i].Nonce + 1; i <= highestNonceInStorer; i++ {
```

Comment:
Change i with j. The i in this for shadows the i from line 339.

Reply:
refactored